lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1234,35 +1235,57 @@ def _on_logging_relation_changed(self, event: HookEvent):
                 charm must update its relation data.
         """
         if isinstance(event, RelationEvent):
-            self._process_logging_relation_changed(event.relation)
+            self._process_logging_relation_changed(event.relation, event)
         else:
             # Upgrade event or other charm-level event
             for relation in self._charm.model.relations[self._relation_name]:
                 self._process_logging_relation_changed(relation)
 
-    def _process_logging_relation_changed(self, relation: Relation):
+    def _process_logging_relation_changed(self, relation: Relation, event: RelationEvent = None):
         """Handle changes in related consumers.
 
-        Anytime there are changes in relations between Loki
-        and its consumers charms, Loki set the `loki_push_api`
-        into the relation data.
-        Besides Loki generates alert rules files based what
-        consumer charms forwards,
+        When there are changes in relations between Loki and its consumers,
+        check whether any changes should be written to the relation data for
+        where to download promtail and/or possible endpoints.
+
+        Additionally, check whether any alert rules are present, and, if so,
+        write them into the Loki container.
 
         Args:
             relation: the `Relation` instance to update.
+            event: an optional event, the type of which is checked. If it is a
+                RelationDepartedEvent no updates are sent to the relation data bag
+
         """
-        if self._charm.unit.is_leader():
-            relation.data[self._charm.app].update(self._promtail_binary_url)
+        if self._charm.unit.is_leader() and event not in [
+            RelationBrokenEvent,
+            RelationDepartedEvent,
+        ]:
+            # Condense the updates so we aren't setting keys to the same values
```

Comment:
Since without this, every `RelationChanged` event will mindlessly set `promtail_binary_url` regardless of whether or not it is already present in the data bag, plus setting endpoints. The changes here are primarily an attempt to "quiet" the relation between `LokiPushApiConsumer` and `LokiPushApiProvider` charms to see if the Juju bug can be isolated

Reply:
I know. But what I was saying for if the relation data contained `foo = bar` and a second invocation of this method again set `foo = bar` into relation data, doesn't Juju guarantee that no `RelationChanged` event will be triggered on the other side of the relation ? My understanding was Juju does make this guarantee. If you have reason to believe that it does not then perhaps we should raise a ticket.

Reply:
Yes, it does. On the hand hand, constantly setting the same data and forcing Juju to respond (plus the invocation time for setting itself [since OF doesn't check the equality either](https://github.com/canonical/operator/blob/6013988d00d68f15ac61a3419bf2a4b2c7f49268/ops/model.py#L1668)) is something we could live without.

But it's the second part of this clause which the changes was made for. If `_endpoints()` **has** changed, we will send an update which is useless to the charm on the other side, and which may have caused some reference in Juju which it was "holding on to". 

It's probably a reasonable change anyway, but it was partly exploratory for a workaround to the Juju issue.